timer/scheduler.go

```diff
@@ -0,0 +1,63 @@
+package timer
+
+import (
+	"time"
+
+	"github.com/nivista/steady/.gen/protos/common"
+	rcron "github.com/robfig/cron/v3"
+)
+
+type cron struct {
+	sched               rcron.Schedule
+	startTime, stopTime time.Time
+	maxExecutions       int
+}
+
+var parser = rcron.NewParser(rcron.Minute | rcron.Hour | rcron.Dom | rcron.Month | rcron.Dow | rcron.SecondOptional | rcron.Descriptor)
+
+// infiniteExecutions indicates a timer should fire until cancelled.
+const infiniteExecutions = 0
+
+func (c cron) schedule(prog Progress, now time.Time) (nextFire time.Time, done bool) {
+	if c.maxExecutions != infiniteExecutions && prog.CompletedExecutions >= c.maxExecutions {
+		done = true
+		return
+	}
+
+	if prog.LastExecution.IsZero() {
+		// subtract one second so sched.Next will return startTime if it is valid
+		nextFire = c.startTime.Add(-1 * time.Second)
+	} else {
+		nextFire = prog.LastExecution
+	}
+
+	for {
+		nextFire = c.sched.Next(nextFire)
+		if c.stopTime.IsZero() && nextFire.After(c.stopTime) {
```

Comment:
I don't understand this condition? If stopTime is zero than won't nextFire always come after stop time?

Reply:
This condition is wrong it's missing a "not". The reason I didn't catch it when running code is beacause IsZero() returns false. The reason IsZero() returns false is that the protos zero value of time is Jan 1 1970, where as the go zero value of time is Jan 1 0001.